### PR TITLE
Added convex optimization for solving relaxed qubo.

### PR DESCRIPTION
**Pull request by mstechly**
TODO:
- polish tests
- make sure the step works.

### CODE REVIEW

**Comment by mstechly** [comment]
I've tested it with a workflow and it works, I've also updated tests.
However, I have the following questions left:
1. Looks like in `solve_qp_relaxation_of_spd_qubo` and `solve_qp_relaxation_of_non_spd_qubo` I don't really need to use `qubo` – these could simply take matrix as an input and I think that would be cleaner.
2. If they took matrix as an input, it would be much easier to write some more elaborate tests. Actually, we could even do property testing there 🤔 
3. Once 1 is resolved, I'll update docstrings and then I think it's ready to go?

CC: @dexter2206

**Comment by dexter2206** [comment]
@mstechly Yes, I think pasing matrices to `solve_qp_relaxation....` is a better approach. 

Some minor things to consider:
- Instead of using indirectly parametrized fixture with boolean flag wouldn't it be cleaner to just use separate fixtures?
- Why are some QUBOs defined as fixtures and others are just placed in the body of test functions? Both approaches are fine, but I think it would be better if we decided to pick one :slightly_smiling_face:

File: src/python/zquantum/qubo/convex_opt.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+def is_matrix_semi_positive_definite(
+    matrix: np.ndarray, epsilon: float = 1e-15
+) -> bool:
+    eigenvalues, _ = np.linalg.eig(matrix)
+    return np.min(eigenvalues) >= epsilon
+
+
+def regularize_relaxed_solution(
+    relaxed_solution: np.ndarray, epsilon: int = 0.5
```
[the remark this answers is not in the document]
Comment (mstechly) [reply src/python/zquantum/qubo/convex_opt.py]: Already noticed and fixed that :)

File: src/python/zquantum/qubo/convex_opt.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    regularized_solution = []
+    for value in relaxed_solution:
+        if value > epsilon and value < 1 - epsilon:
+            regularized_value = 2 * np.arcsin(np.sqrt(value))
+        elif value <= epsilon:
+            regularized_value = 2 * np.arcsin(np.sqrt(epsilon))
+        else:
+            regularized_value = 2 * np.arcsin(np.sqrt(1 - epsilon))
+        regularized_solution.append(regularized_value)
+    return np.array(regularized_solution)
```
[the remark this answers is not in the document]
Comment (mstechly) [reply src/python/zquantum/qubo/convex_opt.py]: Done!

File: steps/relaxed_qubo.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    number_of_trials=10,
+):
+    qubo = load_qubo(qubo)
+    optimizer = create_object(optimizer_specs)
+    solution, optimal_value = solve_qp_relaxation_of_qubo(
+        qubo, optimizer, number_of_trials
+    )
+
+    if regularize_solution:
+        solution = regularize_solution()
```
[the remark this answers is not in the document]
Comment (mstechly) [reply steps/relaxed_qubo.py]: I've refactored it a little bit, WDYT?

File: src/python/zquantum/qubo/convex_opt.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    cost_function = lambda x: x.T @ qubo_matrix @ x
+    final_value = None
+    final_params = None
+
+    for _ in range(number_of_trials):
+        initial_params = np.random.uniform(0.0, 1.0, size=size)
+        optimization_results = optimizer.minimize(
+            cost_function, initial_params, constraints=[linear_constraint]
+        )
+        if final_value is None or final_value < optimization_results.opt_value:
```
[the remark this answers is not in the document]
Comment (mstechly) [reply src/python/zquantum/qubo/convex_opt.py]: Good catch! Fixed

**Review by Guoming-Wang — APPROVED** [review]
Looks good now!

**Review by dexter2206 — APPROVED** [review]

**Pull request merged**